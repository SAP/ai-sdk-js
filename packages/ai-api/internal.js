export * from './dist/internal.js';
//# sourceMappingURL=internal.js.map
