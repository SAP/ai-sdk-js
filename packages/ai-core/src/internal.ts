export * from './client/index.js';
export * from './utils/index.js';
