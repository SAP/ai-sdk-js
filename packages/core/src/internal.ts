export * from './context.js';
export * from './http-client.js';
export * from './openapi-request-builder.js';
export * from './model-types.js';
