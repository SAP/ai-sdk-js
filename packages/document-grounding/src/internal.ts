export * from './client/api/index.js';
