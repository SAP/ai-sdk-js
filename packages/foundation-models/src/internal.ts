export * from './azure-openai/index.js';
