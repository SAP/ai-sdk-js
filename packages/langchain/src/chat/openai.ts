import { CallbackManagerForLLMRun } from '@langchain/core/callbacks/manager';
import { BaseChatModelParams } from '@langchain/core/language_models/chat_models';
import { AIMessage, BaseMessage, ChatMessage, ToolMessage } from '@langchain/core/messages';
import { ChatResult } from '@langchain/core/outputs';
import { StructuredTool } from '@langchain/core/tools';
import { ChatOpenAI, ChatOpenAICallOptions, OpenAIChatInput } from '@langchain/openai';
import { zodToJsonSchema } from 'zod-to-json-schema';
import {
  OpenAiChatAssistantMessage,
  OpenAiChatToolMessage,
  OpenAiClient,
  OpenAiChatCompletionFunction,
  OpenAiChatFunctionCall,
  OpenAiChatMessage,
  OpenAiChatCompletionTool,
  OpenAiChatCompletionOutput,
  OpenAiChatCompletionParameters,
  DeploymentIdConfiguration,
  OpenAiChatModel
} from '@sap-ai-sdk/gen-ai-hub';

/**
 * Input for Text generation for OpenAI GPT.
 */
export interface OpenAIChatModelInterface
  extends Omit<OpenAIChatInput, 'openAIApiKey' | 'streaming'>,
    Omit<OpenAiChatCompletionParameters, 'n' | 'stop' | 'messages' | 'temperature'>,
    BaseChatModelParams {
      /**
       * The deployment ID of the model.
       */
      deploymentId?: string;
      /**
       * The version of the model.
       */
      modelVersion?: string;
    }

/**
 * Input for Text generation for OpenAI GPT.
 */
export type OpenAIChatModelInput = Omit<OpenAIChatInput,
  'frequencyPenalty'
  | 'presencePenalty'
  | 'topP'
  | 'temperature'
  | 'stop'
  | 'n'
  | 'modelName'
  | 'model'
  | 'openAIApiKey'
  | 'streaming'> &
  Omit<OpenAiChatCompletionParameters, 'messages'> &
  BaseChatModelParams &
  DeploymentIdConfiguration &
  {
    /**
     * The name of the model.
     */
    modelName: OpenAiChatModel;
    /**
     * The version of the model.
     */
    modelVersion?: string;
    deploymentId?: string;
  };

/**
 * Chat Call options.
 */
interface OpenAIChatCallOptions
  extends Omit<ChatOpenAICallOptions, 'promptIndex' | 'functions' | 'function_call' | 'tools'> {
  functions?: OpenAiChatCompletionFunction[];
  function_call?: OpenAiChatFunctionCall;
  tools?: OpenAiChatCompletionTool[];
}

/**
 * OpenAI Language Model Wrapper to generate texts.
 */
export class OpenAIChat extends ChatOpenAI implements OpenAIChatModelInterface {
  declare CallOptions: OpenAIChatCallOptions;

  deploymentId?: string;
  modelVersion?: string;
  private btpOpenAIClient: OpenAiClient;

  constructor(fields: OpenAIChatModelInput) {
    const defaultValues = new ChatOpenAI();
    const n = fields.n ?? defaultValues.n;
    const stop = fields.stop ? Array.isArray(fields.stop) ? fields.stop : [fields.stop] : defaultValues.stop;
    const temperature = fields.temperature ?? defaultValues.temperature;
    const frequencyPenalty = fields.frequency_penalty ?? defaultValues.frequencyPenalty;
    const presencePenalty = fields.presence_penalty ?? defaultValues.presencePenalty;
    const topP = fields.top_p ?? defaultValues.topP;

    super({
      ...fields,
      n,
      stop,
      temperature,
      openAIApiKey: 'dummy',
      frequencyPenalty,
      presencePenalty,
      topP
    });

    this.deploymentId = fields.deploymentId;
    this.modelVersion = fields.modelVersion;

    // LLM client
    this.btpOpenAIClient = new OpenAiClient();
  }

  override get callKeys(): (keyof OpenAIChatCallOptions)[] {
    return [
      ...(super.callKeys as (keyof OpenAIChatCallOptions)[]),
      'options',
      'function_call',
      'functions',
      'tools',
      'tool_choice',
      'response_format',
      'seed',
    ];
  }

  override get lc_secrets(): { [key: string]: string } | undefined {
    // overrides default keys as not applicable in BTP
    return {};
  }

  override get lc_aliases(): Record<string, string> {
    // overrides default keys as not applicable in BTP
    return {};
  }

  override async _generate(
    messages: BaseMessage[],
    options: this['CallOptions'],
    runManager?: CallbackManagerForLLMRun,
  ): Promise<ChatResult> {
    function isStructuredToolArray(tools?: unknown[]): tools is StructuredTool[] {
      return tools !== undefined && tools.every((tool) => Array.isArray((tool as StructuredTool).lc_namespace));
    }
    const res = await this.caller.callWithOptions(
      {
        signal: options.signal,
      },
      () =>
        this.btpOpenAIClient.chatCompletion(
          {
            messages: messages.map(this.mapBaseMessageToOpenAIChatMessage.bind(this)),
            deployment_id: this.deploymentId,
            max_tokens: this.maxTokens === -1 ? undefined : this.maxTokens,
            temperature: this.temperature,
            top_p: this.topP,
            logit_bias: this.logitBias,
            n: this.n,
            stop: options?.stop ?? this.stop,
            presence_penalty: this.presencePenalty,
            frequency_penalty: this.frequencyPenalty,
            functions: isStructuredToolArray(options?.functions)
              ? options?.functions.map(this.mapToolToBTPOpenAIFunction.bind(this))
              : options?.functions,
            tools: isStructuredToolArray(options?.tools)
              ? options?.tools.map(this.mapToolToOpenAITool.bind(this))
              : options?.tools,
            tool_choice: options?.tool_choice,
            response_format: options?.response_format,
            seed: options?.seed,
            ...this.modelKwargs,
          },
          {
            modelName: this.modelName ?? this.model,
            deploymentId: this.deploymentId,
            modelVersion: this.modelVersion
          },
        ),
    );

    // currently BTP LLM Proxy for OpenAI doesn't support streaming
    await runManager?.handleLLMNewToken(
      typeof res.choices[0].message.content === 'string' ? res.choices[0].message.content : '',
    );

    return this.mapResponseToChatResult(res);
  }

  /**
   * Maps a LangChain {@link StructuredTool} to {@link OpenAiChatCompletionFunction}.
   * @param tool - Base class for Tools that accept input of any shape defined by a Zod schema.
   * @returns The OpenAI Chat Completion Function.
   */
  protected mapToolToBTPOpenAIFunction(tool: StructuredTool): OpenAiChatCompletionFunction {
    return {
      name: tool.name,
      description: tool.description,
      parameters: zodToJsonSchema(tool.schema),
    };
  }

  /**
   * Maps a LangChain {@link StructuredTool} to {@link OpenAiChatCompletionTool}.
   * @param tool - Base class for Tools that accept input of any shape defined by a Zod schema.
   * @returns The OpenAI Chat Completion Tool.
   */
  protected mapToolToOpenAITool(tool: StructuredTool): OpenAiChatCompletionTool {
    return {
      type: 'function',
      function: {
        name: tool.name,
        description: tool.description,
        parameters: zodToJsonSchema(tool.schema),
      },
    };
  }

  /**
   * Maps a {@link BaseMessage} to OpenAI's Message Role.
   * @param message - The message to map.
   * @returns The OpenAI Message Role.
   */
  protected mapBaseMessageToRole(message: BaseMessage): OpenAiChatMessage['role'] {
    switch (message._getType()) {
      case 'ai':
        return 'assistant';
      case 'human':
        return 'user';
      case 'system':
        return 'system';
      case 'function':
        return 'function';
      case 'tool':
        return 'tool';
      case 'generic':
        return (message as ChatMessage).role as OpenAiChatMessage['role'];
      default:
        throw new Error(`Unknown message type: ${message._getType()}`);
    }
  }

  /**
   * Maps {@link BaseMessage} to OpenAI Messages.
   * @param message - The message to map.
   * @returns The OpenAI Chat Message.
   */
  protected mapBaseMessageToOpenAIChatMessage(message: BaseMessage): OpenAiChatMessage {
    return {
      content: message.content,
      name: message.name,
      role: this.mapBaseMessageToRole(message),
      function_call: message.additional_kwargs.function_call,
      tool_calls: message.additional_kwargs.tool_calls,
      tool_call_id: (message as ToolMessage).tool_call_id,
    } as OpenAiChatMessage;
  }

  /**
   * Maps OpenAI messages to LangChain's {@link ChatResult}.
   * @param res - The OpenAI Chat Completion Output.
   * @returns The LangChain Chat Result.
   */
  protected mapResponseToChatResult(res: OpenAiChatCompletionOutput): ChatResult {
    return {
      generations: res.choices.map((c) => ({
          text: (c.message as OpenAiChatAssistantMessage).content || '',
          message: new AIMessage({
            content: (c.message as OpenAiChatAssistantMessage).content || '',
            additional_kwargs: {
              finish_reason: c.finish_reason,
              index: c.index,
              logprobs: c.logprobs,
              function_call: (c.message as OpenAiChatAssistantMessage).function_call, // add `function_call` parameter
              tool_calls: (c.message as OpenAiChatAssistantMessage).tool_calls,
              tool_call_id: (c.message as OpenAiChatToolMessage).tool_call_id,
            },
          }),
          generationInfo: {
            finish_reason: c.finish_reason,
            index: c.index,
            logprobs: c.logprobs,
            function_call: (c.message as OpenAiChatAssistantMessage).function_call, // add `function_call` parameter
            tool_calls: (c.message as OpenAiChatAssistantMessage).tool_calls,
          },
        })),
      llmOutput: {
        created: res.created,
        id: res.id,
        model: res.model,
        object: res.object,
        tokenUsage: {
          completionTokens: res.usage.completion_tokens,
          promptTokens: res.usage.prompt_tokens,
          totalTokens: res.usage.total_tokens,
        },
      },
    };
  }
}
