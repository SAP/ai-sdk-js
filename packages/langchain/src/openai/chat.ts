import { CallbackManagerForLLMRun } from '@langchain/core/callbacks/manager';
import { BaseMessage } from '@langchain/core/messages';
import type { ChatResult } from '@langchain/core/outputs';
import { AzureOpenAiChatClient as AzureOpenAiChatClientBase } from '@sap-ai-sdk/foundation-models';
import { BaseChatModel } from '@langchain/core/language_models/chat_models';
import { mapLangchainToAiClient, mapOutputToChatResult } from './util.js';
import type {
  AzureOpenAiChatCallOptions,
  AzureOpenAiChatModelParams
} from './types.js';

/**
 * LangChain chat client for Azure OpenAI consumption on SAP BTP.
 */
export class AzureOpenAiChatClient extends BaseChatModel<AzureOpenAiChatCallOptions> {
  temperature?: number | null;
  top_p?: number | null;
  logit_bias?: Record<string, any> | null | undefined;
  user?: string;
  presence_penalty?: number;
  frequency_penalty?: number;
  stop?: string | string[];
  max_tokens?: number;
  private openAiChatClient: AzureOpenAiChatClientBase;

  constructor(fields: AzureOpenAiChatModelParams) {
    super(fields);
    this.openAiChatClient = new AzureOpenAiChatClientBase(fields);
    this.temperature = fields.temperature;
    this.top_p = fields.top_p;
    this.logit_bias = fields.logit_bias;
    this.user = fields.user;
    this.stop = fields.stop;
    this.presence_penalty = fields.presence_penalty;
    this.frequency_penalty = fields.frequency_penalty;
    this.max_tokens = fields.max_tokens;
  }

  _llmType(): string {
    return 'azure_openai';
  }

  override async _generate(
    messages: BaseMessage[],
    options: typeof this.ParsedCallOptions,
    runManager?: CallbackManagerForLLMRun
  ): Promise<ChatResult> {
    const res = await this.caller.callWithOptions(
      {
        signal: options.signal
      },
      () =>
        this.openAiChatClient.run(
          mapLangchainToAiClient(this, messages, options),
          options.requestConfig
        )
    );

    const content = res.getContent();

    // we currently do not support streaming
    await runManager?.handleLLMNewToken(
      typeof content === 'string' ? content : ''
    );

    return mapOutputToChatResult(res.data);
  }
}
