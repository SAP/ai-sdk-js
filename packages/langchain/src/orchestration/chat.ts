import { BaseChatModel } from '@langchain/core/language_models/chat_models';
import { OrchestrationClient as OrchestrationClientBase } from '@sap-ai-sdk/orchestration';
import { AsyncCaller } from '@langchain/core/utils/async_caller';
import { resilience } from '@sap-cloud-sdk/resilience';
import {
  isTemplate,
  mapLangchainMessagesToOrchestrationMessages,
  mapOutputToChatResult
} from './util.js';
import type { CustomRequestConfig } from '@sap-ai-sdk/core';
import type { OrchestrationMessageChunk } from './orchestration-message-chunk.js';
import type { ChatResult } from '@langchain/core/outputs';
import type { OrchestrationModuleConfig } from '@sap-ai-sdk/orchestration';
import type { BaseChatModelParams } from '@langchain/core/language_models/chat_models';
import type { ResourceGroupConfig } from '@sap-ai-sdk/ai-api';
import type { BaseMessage } from '@langchain/core/messages';
import type { CallbackManagerForLLMRun } from '@langchain/core/callbacks/manager';
import type { OrchestrationCallOptions } from './types.js';
import type { HttpDestinationOrFetchOptions } from '@sap-cloud-sdk/connectivity';

// TODO: Update all docs

/**
 * LangChain chat client for Azure OpenAI consumption on SAP BTP.
 */
export class OrchestrationClient extends BaseChatModel<
  OrchestrationCallOptions,
  OrchestrationMessageChunk
> {
  constructor(
    // Omit streaming until supported
    public orchestrationConfig: Omit<OrchestrationModuleConfig, 'streaming'>,
    public langchainOptions: BaseChatModelParams = {},
    public deploymentConfig?: ResourceGroupConfig,
    public destination?: HttpDestinationOrFetchOptions
  ) {
    super(langchainOptions);
  }

  _llmType(): string {
    return 'orchestration';
  }

  /**
   * Decisions:
   * bind only supports ParsedCallOptions, we don't support arbitrary LLM options, only tool calls & default BaseLanguageModelCallOptions, e.g. stop ✅
   * this aligns with other vendors' client designs (e.g. openai, google) ✅
   * top of the array (array[array.length - 1]) contains the current message, everything before then is history.
   * Module results are part of our own message type, which extends AI Message to work with all other langchain functionality.
   *
   * For timeout, we need to apply our own middleware, it is not handled by langchain. ✅.
   */

  override async _generate(
    messages: BaseMessage[],
    options: typeof this.ParsedCallOptions,
    runManager?: CallbackManagerForLLMRun
  ): Promise<ChatResult> {
    let caller = this.caller;
    if (options.maxConcurrency) {
      const { maxConcurrency, maxRetries, onFailedAttempt } =
        this.langchainOptions;
      caller = new AsyncCaller({
        maxConcurrency: maxConcurrency ?? options.maxConcurrency,
        maxRetries,
        onFailedAttempt
      });
    }
    const res = await caller.callWithOptions(
      {
        signal: options.signal
      },
      () => {
        // consider this.tools & this.stop property, merge it with template orchestration config
        const { inputParams } = options;
        const mergedOrchestrationConfig =
          this.mergeOrchestrationConfig(options);
        const orchestrationClient = new OrchestrationClientBase(
          mergedOrchestrationConfig,
          this.deploymentConfig,
          this.destination
        );
        const messagesHistory =
          mapLangchainMessagesToOrchestrationMessages(messages);
        const customRequestConfig: CustomRequestConfig = {
          ...options.customRequestConfig,
          middleware: resilience({ timeout: options.timeout })
        };
        return orchestrationClient.chatCompletion(
          {
            // how to handle tools here? doesn't really exist as input in orchestration as message history
            // make template a call option, to merge it ??
            messagesHistory,
            inputParams
          },
          customRequestConfig
        );
      }
    );

    const content = res.getContent();

    // we currently do not support streaming
    await runManager?.handleLLMNewToken(
      typeof content === 'string' ? content : ''
    );

    return mapOutputToChatResult(res.data);
  }

  private mergeOrchestrationConfig(
    options: typeof this.ParsedCallOptions
  ): OrchestrationModuleConfig {
    const { tools = [], stop = [] } = options;
    return {
      ...this.orchestrationConfig,
      llm: {
        ...this.orchestrationConfig.llm,
        model_params: {
          ...this.orchestrationConfig.llm.model_params,
          ...(stop.length && {
            stop: [
              ...(this.orchestrationConfig.llm.model_params?.stop || []),
              ...stop
            ]
          })
        }
      },
      templating: {
        ...this.orchestrationConfig.templating,
        ...(isTemplate(this.orchestrationConfig.templating) &&
          tools.length && {
            tools: [
              ...(this.orchestrationConfig.templating.tools || []),
              ...tools
            ]
          })
      }
    };
  }
}
