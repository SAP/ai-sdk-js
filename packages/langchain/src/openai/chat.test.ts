import { tool } from '@langchain/core/tools';
import { jest } from '@jest/globals';
import { zodToJsonSchema } from 'zod-to-json-schema';
import nock from 'nock';
import { z } from 'zod';
import { apiVersion } from '@sap-ai-sdk/foundation-models/internal.js';
import {
  mockClientCredentialsGrantCall,
  mockDeploymentsList,
  mockInference
} from '../../../../test-util/mock-http.js';
import { AzureOpenAiChatClient } from './chat.js';

describe('Chat client', () => {
  let client: AzureOpenAiChatClient;
  const endpoint = {
    url: 'inference/deployments/1234/chat/completions',
    apiVersion
  };

  beforeEach(async () => {
    client = new AzureOpenAiChatClient({ modelName: 'gpt-4o' });
    mockClientCredentialsGrantCall();
    mockDeploymentsList(
      {
        scenarioId: 'foundation-models',
        resourceGroup: 'default',
        executableId: 'azure-openai'
      },
      { id: '1234', model: { name: 'gpt-4o', version: 'latest' } }
    );
  });

  afterEach(() => {
    nock.cleanAll();
  });

  describe('bindTools', () => {
    const toolResponse = {
      data: {
        choices: [
          {
            message: {
              role: 'assistant',
              content: null,
              tool_calls: [
                {
                  id: '1',
                  type: 'function',
                  function: {
                    name: 'add',
                    arguments: '{ "a": 1, "b": 2 }'
                  }
                }
              ]
            },
            index: 0
          }
        ]
      }
    };
    const addNumbersSchema = z
      .object({
        a: z.number().describe('The first number to be added.'),
        b: z.number().describe('The second number to be added.')
      })
      .strict();
    const myTool = tool(({ a, b }) => a + b, {
      name: 'add',
      description: 'Add two numbers',
      schema: addNumbersSchema
    });

    it('should bind a tool with strict set to true if defined in kwargs', async () => {
      const clientSpy = jest.spyOn(client['openAiChatClient'], 'run');
      mockInference(
        {
          data: {
            messages: [
              {
                role: 'user' as const,
                content: 'What is 1 + 2?'
              }
            ],
            tools: [
              {
                type: 'function',
                function: {
                  name: 'add',
                  description: 'Add two numbers',
                  parameters: zodToJsonSchema(addNumbersSchema),
                  strict: true // Will be tested
                }
              }
            ]
          }
        },
        toolResponse,
        endpoint
      );
      await client
        .bindTools([myTool], { strict: true })
        .invoke('What is 1 + 2?');
      expect(clientSpy).toHaveBeenCalledTimes(1);
    });

    it('should bind a tool with strict set to false if defined in kwargs', async () => {
      const clientSpy = jest.spyOn(client['openAiChatClient'], 'run');
      mockInference(
        {
          data: {
            messages: [
              {
                role: 'user' as const,
                content: 'What is 1 + 2?'
              }
            ],
            tools: [
              {
                type: 'function',
                function: {
                  name: 'add',
                  description: 'Add two numbers',
                  parameters: zodToJsonSchema(addNumbersSchema),
                  strict: false // Will be tested
                }
              }
            ]
          }
        },
        toolResponse,
        endpoint
      );
      await client
        .bindTools([myTool], { strict: false })
        .invoke('What is 1 + 2?');
      expect(clientSpy).toHaveBeenCalledTimes(1);
    });

    it('should bind a tool with undefined strict if not defined in kwargs', async () => {
      const clientSpy = jest.spyOn(client['openAiChatClient'], 'run');
      mockInference(
        {
          data: {
            messages: [
              {
                role: 'user' as const,
                content: 'What is 1 + 2?'
              }
            ],
            tools: [
              {
                type: 'function',
                function: {
                  name: 'add',
                  description: 'Add two numbers',
                  parameters: zodToJsonSchema(addNumbersSchema),
                  strict: undefined // Will be tested
                }
              }
            ]
          }
        },
        toolResponse,
        endpoint
      );
      await client.bindTools([myTool]).invoke('What is 1 + 2?');
      expect(clientSpy).toHaveBeenCalledTimes(1);
    });

    it('should bind a tool with strict set to true if defined by supportsStrictToolCalling', async () => {
      client.supportsStrictToolCalling = true;
      const clientSpy = jest.spyOn(client['openAiChatClient'], 'run');
      mockInference(
        {
          data: {
            messages: [
              {
                role: 'user' as const,
                content: 'What is 1 + 2?'
              }
            ],
            tools: [
              {
                type: 'function',
                function: {
                  name: 'add',
                  description: 'Add two numbers',
                  parameters: zodToJsonSchema(addNumbersSchema),
                  strict: true // Will be tested
                }
              }
            ]
          }
        },
        toolResponse,
        endpoint
      );
      await client.bindTools([myTool]).invoke('What is 1 + 2?');
      expect(clientSpy).toHaveBeenCalledTimes(1);
    });
  });
});
