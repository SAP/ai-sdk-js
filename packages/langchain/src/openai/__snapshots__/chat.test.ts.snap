// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat client streaming supports streaming responses 1`] = `
{
  "id": [
    "langchain_core",
    "messages",
    "AIMessageChunk",
  ],
  "kwargs": {
    "additional_kwargs": {},
    "content": "The capital of France is Paris.",
    "id": undefined,
    "invalid_tool_calls": [],
    "response_metadata": {
      "completion": 0,
      "created": 1730125149,
      "finish_reason": "stop",
      "id": "chatcmpl-ANKsHIdjvozwuOGpGI6rygvwSJH0I",
      "index": 0,
      "model_name": "gpt-4o",
      "prompt": 0,
      "system_fingerprint": "fp_808245b034",
      "token_usage": {
        "completion_tokens": 7,
        "prompt_tokens": 14,
        "total_tokens": 21,
      },
    },
    "tool_call_chunks": [],
    "tool_calls": [],
    "usage_metadata": {
      "input_tokens": 14,
      "output_tokens": 7,
      "total_tokens": 21,
    },
  },
  "lc": 1,
  "type": "constructor",
}
`;

exports[`Chat client streaming supports streaming responses with tool calls 1`] = `
{
  "id": [
    "langchain_core",
    "messages",
    "AIMessageChunk",
  ],
  "kwargs": {
    "additional_kwargs": {},
    "content": "",
    "id": undefined,
    "invalid_tool_calls": [],
    "response_metadata": {
      "completion": 0,
      "created": 1750673370,
      "finish_reason": "tool_calls",
      "id": "chatcmpl-BlYPCAKjvl7PBJxuZb352G12WIQK5",
      "index": 0,
      "model_name": "gpt-4o-2024-08-06",
      "prompt": 0,
      "system_fingerprint": "fp_ee1d74bde0",
      "token_usage": {
        "completion_tokens": 18,
        "completion_tokens_details": {
          "accepted_prediction_tokens": 0,
          "audio_tokens": 0,
          "reasoning_tokens": 0,
          "rejected_prediction_tokens": 0,
        },
        "prompt_tokens": 52,
        "prompt_tokens_details": {
          "audio_tokens": 0,
          "cached_tokens": 0,
        },
        "total_tokens": 70,
      },
    },
    "tool_call_chunks": [
      {
        "args": "{"a":1,"b":2}",
        "id": "call_De0ejo2G1gknErC39DDH2JpS",
        "index": 0,
        "name": "add",
        "type": "tool_call_chunk",
      },
    ],
    "tool_calls": [
      {
        "args": {
          "a": 1,
          "b": 2,
        },
        "id": "call_De0ejo2G1gknErC39DDH2JpS",
        "name": "add",
        "type": "tool_call",
      },
    ],
    "usage_metadata": {
      "input_tokens": 52,
      "output_tokens": 18,
      "total_tokens": 70,
    },
  },
  "lc": 1,
  "type": "constructor",
}
`;
